### ISSUE TITLE
delete for MutableList

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1455]
Is there a way to manipulate mutable lists without making a new copy? For instance:
```m2
A = new MutableList from {1,2,3};
append(A, 10) -- MutableList{...4...}
peek A -- MutableList{1, 2, 3}
```
For comparison, mutable hash tables can be manipulated in place:
```m2
B = new MutableHashTable from {a => 2}
B#b = 4
peek B
-- MutableHashTable{a => 2}
--                  b => 4
remove(B, b)
peek B -- MutableHashTable{a => 2}
```

I assume no copies are being made here, or at least I hope so!

**Comment by DanGrayson** [comment]
Yes:
```m2
Macaulay2, version 1.16
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : A = new MutableList

o1 = MutableList{}

o1 : MutableList

i2 : A#0=x

o2 = x

o2 : Symbol

i3 : A#1=y

o3 = y

o3 : Symbol

i4 : #A

o4 = 2

i5 : A##A=z -- here's how to append

o5 = z

o5 : Symbol

i6 : peek A

o6 = MutableList{x, y, z}
```

**Comment by DanGrayson** [comment]
PS: yes, when modifying a mutable hash table, no copies are made.

**Comment by mahrud** [comment]
Ah, thank you. How bad would it be to modify `append` and `prepend` to do this for mutable lists?

**Comment by mahrud** [comment]
Reopening to remember the bug with `delete` on a MutableList.

**Comment by mahrud** [comment]
```m2
i1 : L = new MutableList from {"A", "B"}

o1 = MutableList{...2...}

o1 : MutableList

i2 : mutable L

o2 = true

i3 : delete("A", L)

o3 = MutableList{...1...}

o3 : MutableList

i4 : mutable oo

o4 = false

i5 : delete(L, 1)

o5 = {0}

o5 : List
```

**Comment by mahrud** [comment]
`drop` for mutable lists allocates new memory instead of just dropping those elements. @d-torrance I think this should be an easy fix, right?

**Comment by d-torrance** [comment]
Yeah, that should be pretty easy.  Is that the desired behavior, though?  Different behavior for mutable and immutable lists?

We do have `remove`, which does modify the list itself.

**Comment by d-torrance** [comment]
That makes sense to me!  I guess why use a mutable list if you don't want to mutate it?

So in this issue so far, it looks like the following have been discussed as candidates for this behavior change:

* `append`
* `delete`
* `drop`
* `prepend`

Are there any others?  Maybe `join`? `insert`?

**Comment by mahrud** [comment]
I remembered why this is hard: it's because mutable lists are not linked lists ... but once they are, all of these should be trivial to implement (also slices!)

See #1608.

**Comment by mahrud** [comment]
To answer your question, I think `append` and `join` are easy to implement with the current implementation, but with the rest, it's possible that sometimes allocating a new mutable list directly with the modifying data is actually faster than first making a copy of the mutable list and then modifying it.
